pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConnectionHandler.java

```diff
@@ -81,31 +91,44 @@ protected void grabCnx() {
             } else {
                 cnxFuture = state.client.getConnection(state.topic); //
             }
-            cnxFuture.thenAccept(cnx -> connection.connectionOpened(cnx)) //
-                    .exceptionally(this::handleConnectionError);
+            cnxFuture.whenComplete((cnx, e) -> {
+                try {
+                    final CompletableFuture<Void> future = connection.handleNewConnection(cnx, convertThrowable(e));
+                    if (e != null) {
+                        State state = this.state.getState();
+                        if (state == State.Uninitialized || state == State.Connecting || state == State.Ready) {
+                            reconnectLater(e);
```

Comment:
Although generally the timer will schedule the reconnect after `handleNewConnection` finished, I'm thinking of building an explicit happens-before order bwteen these two calls.

What if we retain `connectionOpened` and `connectionFailed` but change them to async and handle them separately in `thenAccept` and `exceptionally`? It seems easier to reason.

Reply:
That is, `reconnectLater` calls `CLIENT_CNX_UPDATER.set(this, null);` while `handleNewConnection` calls `CLIENT_CNX_UPDATER.set(this, clientCnx);` on normal case. It's not quite intuitive but depend on implement details so I'm concerned that it can be brittle.

Reply:
Good point. I will try this way.

Reply:
Addressed.